sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ReplicatedResourceClient.java

```diff
@@ -158,37 +163,63 @@ public Mono<StoreResponse> invokeAsync(RxDocumentServiceRequest request,
     private Mono<StoreResponse> getStoreResponseMonoWithSpeculation(RxDocumentServiceRequest request, Quadruple<Boolean, Boolean, Duration, Integer> forceRefreshAndTimeout) {
         CosmosEndToEndOperationLatencyPolicyConfig config = request.requestContext.getEndToEndOperationLatencyPolicyConfig();
         AvailabilityStrategy strategy = config.getAvailabilityStrategy();
-        List<Mono<StoreResponse>> monoList = new ArrayList<>();
-        List<RxDocumentServiceRequest> requestList = new ArrayList<>();
 
-        if (strategy instanceof ThresholdBasedAvailabilityStrategy) {
-            List<URI> effectiveEndpoints = getApplicableEndPoints(request);
-            if (effectiveEndpoints != null) {
-                effectiveEndpoints
-                    .forEach(locationURI -> {
-                        if (locationURI != null) {
-                            RxDocumentServiceRequest newRequest = request.clone();
-                            newRequest.requestContext.routeToLocation(locationURI);
-                            requestList.add(newRequest);
-                            if (monoList.isEmpty()) {
-                                monoList.add(getStoreResponseMono(newRequest, forceRefreshAndTimeout));
-                            } else {
-                                monoList.add(getStoreResponseMono(newRequest, forceRefreshAndTimeout)
-                                    .delaySubscription(((ThresholdBasedAvailabilityStrategy) strategy).getThreshold()
-                                        .plus(((ThresholdBasedAvailabilityStrategy) strategy)
-                                            .getThresholdStep().multipliedBy(monoList.size() - 1))));
-                            }
-                        }
-                    });
-            }
+        if (!(strategy instanceof ThresholdBasedAvailabilityStrategy)) {
+            // If no threshold based availability strategy is defined just return original Mono
+            return getStoreResponseMono(request, forceRefreshAndTimeout);
         }
 
-        // If the above conditions are not met, then we will just return the original request
-        if (monoList.isEmpty()) {
-            monoList.add(getStoreResponseMono(request, forceRefreshAndTimeout));
+        List<Mono<StoreResponse>> monoList = new ArrayList<>();
+        List<RxDocumentServiceRequest> requestList = new ArrayList<>();
+        final List<URI> effectiveEndpoints = getApplicableEndPoints(request);
```

Comment:
While we are here, I think it's worth double-checking that `getApplicableEndpoints()` do return the endpoints in the correct order as specified in the preferred regions list. (since WMT reported that they were seeing it other way around)

Reply:
Good catch - this is a bug right now - the read locations are not ordered by preference
Which also means the availability strategy retries in different order

Reply:
never mind - it is always ordered correctly

Reply:
No need to change anything - all endpoints are automatically ordered by preference in LocationCache. Renamed variables to make this more clear